File: examples/cpp/chanadd/chanadd.cpp
//=============================================================================
//
// chanadd.cpp
//
// Save the list of database channels to a file, get data from the indicated
// channel and increase it by a given amount.
//
// This is a sample C++ program that illustrates how to connect to the GX
// developer environment from a stand-alone program.  In this case, there
// the following basic steps:
//
//   1. Get (create) a GX Object handle.
//
//   2. Open existing database
//
//   3. Save the list of database channels to a file
//
//   4. Get the channel data
//
//   5. Increase the data by a given amount
//
//   6. Save the changed data back to the channel
//
//   7. No need to destroy any created GX objects (handles): their destructors are
//      automatically called on scope exit
//
// This example also shows how to get the actual data from a channel
//
//=============================================================================

#define PROGRAM L"CHANADD"
#define VERSION L"v1.0"
#define DEFINIT L"Save the list of database channels, increase channel data by a given amount"
#define UTF8SYM L"复制数据专复制数据专复制数据专复制数据专复制数据专复制数据"


#include "stdafx.h"


// standard file library
#include <fstream>


// sample function declaration
void ProcessData(gx_string_type& databaseName, gx_string_type& channelName, double addValue);
void ModifyVAChannel(GXDBPtr db, int32_t channelSymbol, double addValue);
void ModifyVVChannel(GXDBPtr db, int32_t channelSymbol, double addValue);


// unicode to utf8 conversion
void UNICODE_TO_UTF8(const wchar_t* pwUnicode, char* pcUTF8, long lUTF8Size)
{
   long lConv;
   lConv = WideCharToMultiByte(CP_UTF8, 0, pwUnicode, (int)wcslen(pwUnicode), pcUTF8, lUTF8Size, NULL, NULL);
   pcUTF8[max(0, min(lUTF8Size - 1, lConv))] = '\0';
}


// entry point
int wmain(int nargs, TCHAR** args)
{
   try
   {
      // greetings
      wchar_t* buffer = _wgetcwd(NULL, 0);
      wprintf(PROGRAM L" " VERSION L" - " DEFINIT L"\n\nWorking directory = %s\n\n", buffer);
      free(buffer);

      // check script parameters
      if (nargs < 4) throw "Script parameters ERROR - not all required parameters were provided\n";

      // print script parameters
      printf("Parameters:\n");
      for (int i = 0; i < nargs; i++)
         wprintf(L"  Argv[%d] = %s\n", i, args[i]);

      // run an example
      gx_string_type databaseName = args[1];
      gx_string_type channelName = args[2];
      double addValue = _wtof(args[3]);
      ProcessData(databaseName, channelName, addValue);

      wprintf(L"\nDatabase was processed successfully\n");
   }

   // GX API error
   catch (GXAPIError error)
   {
      wprintf(error.message().c_str());
   }

   // other exceptions
   catch (std::exception e)
   {
      printf(e.what());
   }

   catch (char* p)
   {
      printf(p);
   }

   catch (gx_string_type error)
   {
      wprintf(error.c_str());
   }

   catch (...)
   {
      printf("Fatal error");
   }

   // successfully completed
   return 0;
}


// sample function definition
void ProcessData(gx_string_type& databaseName, gx_string_type& channelName, double addValue)
{
   // get (create) a GX Object handle
   GXContextPtr ctx = GXContext::create(std::wstring(L"ChanAdd"), std::wstring(L"Geosoft Inc."));

   try
   {
      // open database
      // note that all databases created by normal Geosoft users will have
      // user name "SUPER" with password "". Third-party or custom applications
      // may create databases that require a user-name and password, in which
      // case you require these values to open the database.
      gx_string_type userName = gx_string_type(gx_string_literal("SUPER"));
      gx_string_type password = gx_string_type(gx_string_literal(""));
      GXDBPtr db = GXDB::open(databaseName, userName, password);
      wprintf(L"Database '%s' opened successfully", databaseName.c_str());

      // get list of channels
      GXLSTPtr channelsList = GXLST::create(4096);
      db->chan_lst(channelsList);

      // open file to export list of channels
      std::ofstream channelsListFile;
      gx_string_type channelsListFileName = databaseName + gx_string_literal(".channel_list.utf8.txt");
      channelsListFile.open(channelsListFileName, std::ios_base::trunc | std::ios_base::binary);

      // file with the list of channels is in UTF8
      char utf8_signature[] = "\xEF\xBB\xBF";
      channelsListFile << utf8_signature;

      // save the list of channels to a file
      for (int32_t i = 0; i < channelsList->size(); i++)
      {
         gx_string_type name;
         int32_t symb;
         channelsList->gt_symb_item(i, name, symb);

         gx_string_type text =
            gx_string_literal("Channel[") + std::to_wstring(i) + gx_string_literal("]: ") +
            gx_string_literal("symbol = ") + std::to_wstring(symb) +
            gx_string_literal("   name = '") + name + gx_string_literal("'\r\n");

         //convert to UTF8
         char utf8[1024];
         UNICODE_TO_UTF8(text.c_str(), utf8, sizeof(utf8));

         //save channel info
         channelsListFile << utf8;
      }
      wprintf(L"\nList of channels exported to file '%s'\n", channelsListFileName.c_str());

      // print selected lines
      int32_t nlines = db->count_sel_lines();
      wprintf(L"Number of selected lines to process: %d\n", nlines);
      for (int32_t curLineSymb = db->first_sel_line(); db->is_line_valid(curLineSymb); curLineSymb = db->next_sel_line(curLineSymb))
      {
         gx_string_type name;
         db->get_symb_name(curLineSymb, name);

         gx_string_type text =
            gx_string_literal("  Line '") + name + gx_string_literal("' ") +
            gx_string_literal(" symbol = ") + std::to_wstring(curLineSymb) + gx_string_literal("\n");
         wprintf(text.c_str());
      }

      // find the indicated channel
      int32_t channelSymbol = db->find_chan(channelName);
      if (channelSymbol == NULLSYMB)
         throw gx_string_type(gx_string_literal("EXCEPTION: Channel '")) + channelName + gx_string_type(gx_string_literal("' not found in database"));

      // lock the channel for modification
      db->lock_symb(channelSymbol, DB_LOCK_READWRITE, DB_WAIT_INFINITY);

      // modify the channel (just, as an example, add value to all its elements)
      int32_t nx = db->get_col_va(channelSymbol);
      if (nx > 1)
      {
         // VA channel
         wprintf(L"VA channel '%s' found with %d columns", channelName.c_str(), nx);
         ModifyVAChannel(db, channelSymbol, addValue);
      }
      else
      {
         // VV channel
         wprintf(L"VV channel '%s' found", channelName.c_str());
         ModifyVVChannel(db, channelSymbol, addValue);
      }
   } catch (GXExit)
   {
      printf("Exited");
   } catch (GXCancel)
   {
      printf("Cancelled");
   } catch (GXError& e)
   {
      wprintf(e.message().c_str());
   }
}


// modify VA channel (add some value, as an example)
void ModifyVAChannel(GXDBPtr db, int32_t channelSymbol, double addValue)
{
   // create VV to hold the channel data
   GXVAPtr va = GXVA::create(GEO_VAR::GS_REAL, 0, db->get_col_va(channelSymbol));

   // loop through all lines to process the channel data
   for (int32_t curLineSymb = db->first_sel_line(); db->is_line_valid(curLineSymb); curLineSymb = db->next_sel_line(curLineSymb))
   {
      //read all data from the channel in the current line
      db->get_chan_va(curLineSymb, channelSymbol, va);

      //number of rows and columns for the channel
      int32_t nrows = va->len();
      int32_t ncols = va->col();

      //get the retrieved data as an array
      std::vector<double> channelArray = va->get_array<double>(0, 0, nrows, ncols);

      //modify data (add some value, as an example)
      for (int i = 0; i < nrows; i++)
         for (int j = 0; j < ncols; j++)
            channelArray[i * ncols + j] += addValue;

      //put the array back into vv
      va->set_array(0, 0, nrows, ncols, channelArray);

      //write modified va back to the channel
      db->put_chan_va(curLineSymb, channelSymbol, va);
   }
}


// modify VV channel (add some value, as an example)
void ModifyVVChannel(GXDBPtr db, int32_t channelSymbol, double addValue)
{
   // create VV to hold the channel data
   GXVVPtr vv = GXVV::create(GEO_VAR::GS_REAL, 16386);

   // loop through all lines to process the channel data
   for (int32_t curLineSymb = db->first_sel_line(); db->is_line_valid(curLineSymb); curLineSymb = db->next_sel_line(curLineSymb))
   {
      //read all data from the channel in the current line
      db->get_chan_vv(curLineSymb, channelSymbol, vv);

      //number of rows for the channel
      int32_t nrows = vv->length();

      //get the retrieved data as an array
      std::vector<double> channelArray = vv->get_data<double>(0, nrows);

      //modify data (add some value, as an example)
      for (int i = 0; i < nrows; i++)
         channelArray[i] += addValue;

      //put the array back into vv
      vv->set_data(0, channelArray);

      //write modified va back to the channel
      db->put_chan_vv(curLineSymb, channelSymbol, vv);
   }
}

File: examples/Python/chanadd/chanadd.py
#!python3.3
'''
Add a constant value to a channel on all selected lines.

This is a sample Python program that illustrates how to connect to the GX
developer environment from a stand-alone program.  In this case, there
the following basic steps:

   1. Get (create) a GX Object handle.

   2. Open existing database

   3. Save the list of database channels to a file

   4. Get the channel data

   5. Increase the data by a given amount

   6. Save the changed data back to the channel
'''
import os
import sys
import geosoft.gx as gx
import numpy as np
import argparse as argp

def process_database(database_name, channel_name, add_value):
    '''
    Process all selected lines in a database by adding a constant 
    value to a channel. Works with 'regular' or array channels.
    
    The data is processed in-place.
    '''
    # get (create) a GX context
    ctx = gx.GXContext.create('Channel Add', 'Geosoft Inc.')

    # open database
    # note that all databases created by normal Geosoft users will have
    # user name "SUPER" with password "". Third-party or custom applications
    # may create databases that require a user-name and password, in which
    # case you require these values to open the database.
    username = "SUPER"
    password = ""
    db = gx.GXDB.open(ctx, database_name, username, password)
    print("Database '" + database_name + "' opened successfully")

    # get list of channels
    channels_list = gx.GXLST.create(ctx, 4096)
    db.chan_lst(channels_list)

    # open file to export list of channels
    channels_list_file_name = database_name + ".channel_list.txt"
    channels_list_file = open(channels_list_file_name, 'w')
  
    # save the list of channels to a file
    for i in range(channels_list.size()):
        name = gx.str_ref()
        symb = gx.int_ref()
        channels_list.gt_symb_item(i, name, symb)
        line = "Channel[" + str(i) + "]: symbol = " + str(symb.value) + "   name = '" + name.value + "'"
        channels_list_file.write(line + "\n")
    print("List of channels exported to file '" + channels_list_file_name + "'")

    # print selected lines
    nlines = db.count_sel_lines()
    print("Number of selected lines to process: " + str(nlines))
    cur_line_symb = db.first_sel_line()
    while db.is_line_valid(cur_line_symb):
        name = gx.str_ref()
        db.get_symb_name(cur_line_symb, name)
        print("  Line '" + name.value + "' symbol = " + str(cur_line_symb))
        cur_line_symb = db.next_sel_line(cur_line_symb)

    # find the indicated channel
    channel_symbol = db.find_chan(channel_name)
    print( channel_symbol)
    if channel_symbol == gx.NULLSYMB:
        raise Exception("EXCEPTION: Channel '" + channel_name + "' not found in database")

    # lock the channel for modification
    db.lock_symb(channel_symbol, gx.DB_LOCK_READWRITE, gx.DB_WAIT_INFINITY)

    # modify the channel (just, as an example, add value to all its elements)
    nx = db.get_col_va(channel_symbol)
    if nx > 1:
        # VA channel
        print("VA channel '" + channel_name + "' found with " + str(nx) + " columns")
        modify_va_channel(ctx, db, channel_symbol, add_value)
    else:
        # VV channe
        print("VV channel '" + channel_name + "' found")
        modify_vv_channel(ctx, db, channel_symbol, add_value)


def modify_va_channel(ctx, db, channel_symbol, add_value):
    '''
    Iterate over selected lines, and add value. This is the
    version used for VA (array) channels.
    '''
    # create an object to hold the channel data
    va = gx.GXVA.create(ctx, gx.GS_REAL, 0, db.get_col_va(channel_symbol))

    # loop through all lines to process the channel data
    cur_line_symb = db.first_sel_line()
    while db.is_line_valid(cur_line_symb):

        # read all data from the channel in the current line
        db.get_chan_va(cur_line_symb, channel_symbol, va)

        # number of rows and columns for the channel
        nrows = va.len()
        ncols = va.col()

        # get the retrieved data as a 1d array, and reshape the array to 2D (va.get_array_np should have done this!)
        channel_array = va.get_array_np(0, 0, nrows, ncols, np.float64)
        channel_array.shape = (nrows, ncols)

        # add value
        channel_array += add_value

        # put the array back into va
        va.set_array_np(0, 0, channel_array)

        # write modified va back to the channel
        db.put_chan_va(cur_line_symb, channel_symbol, va)

        # get the next selected line
        cur_line_symb = db.next_sel_line(cur_line_symb)


def modify_vv_channel(ctx, db, channel_symbol, add_value):
    '''
    Iterate over selected lines, and add value. This is the
    version used for VV channels.
    '''
    #create an object to hold the channel data
    vv = gx.GXVV.create(ctx, gx.GS_REAL, 0)
    
    # loop through all lines to process the channel data
    cur_line_symb = db.first_sel_line()
    while db.is_line_valid(cur_line_symb):

        # read all data from the channel in the current line
        db.get_chan_vv(cur_line_symb, channel_symbol, vv)

        # number of rows for the channel
        nrows = vv.length()

        # get the retrieved data as an array
        channel_array = vv.get_data_np(0, nrows, np.float64)

        # modify data (add some value, as an example)
        channel_array += add_value

        # put the array back into vv
        vv.set_data_np(0, channel_array)

        # write modified va back to the channel
        db.put_chan_vv(cur_line_symb, channel_symbol, vv)

        # get the next selected line
        cur_line_symb = db.next_sel_line(cur_line_symb)

   
  
if __name__ == "__main__":

    # GEOSOFT Environment variable should contain a path with geodist.dll
    print("GX_GEOSOFT_BIN_PATH = {}".format(os.getenv("GX_GEOSOFT_BIN_PATH")))
    print('Working directory = ' + os.path.abspath(os.curdir))

    # get command line parameters
    parser = argp.ArgumentParser(description="Add a constant to a Geosoft database channel")
    parser.add_argument("sDB", help="Geosoft database")
    parser.add_argument("sCh", help="channel to process")
    parser.add_argument("-v","--value", type=float, default=1.0, help="value to add, default is 1.0")
    args = parser.parse_args()
    print("chanadd - no copyright\n")

    # echo parameters
    print("Database = "+args.sDB)
    print("Channel = "+args.sCh)
    print("Value = {}".format(args.value))

    process_database(args.sDB, args.sCh, args.value)
